src/modules/powerrename/lib/Settings.cpp

```diff
@@ -52,34 +57,208 @@ namespace
+    {
+        Reset();
+        return false;
+    }
+    // Go backwards to consume latest items first.
+    long idx = (pushIdx + size - nextIdx) % size;
+    if (items[idx].empty())
+    {
+        Reset();
+        return false;
     }
+    data = items[idx];
+    ++nextIdx;
+    return true;
 }
 
-typedef int (CALLBACK* MRUCMPPROC)(LPCWSTR, LPCWSTR);
+void MRUListHandler::Reset()
+{
+    nextIdx = 1;
+}
 
-typedef struct {
-    DWORD      cbSize;
-    UINT       uMax;
-    UINT       fFlags;
-    HKEY       hKey;
-    LPCTSTR    lpszSubKey;
-    MRUCMPPROC lpfnCompare;
-} MRUINFO;
+void MRUListHandler::Load()
+{
+    if (!std::filesystem::exists(jsonFilePath))
+    {
+        MigrateFromRegistry();
 
-typedef HANDLE (*CreateMRUListFn)(MRUINFO* pmi);
-typedef int (*AddMRUStringFn)(HANDLE hMRU, LPCWSTR data);
-typedef int (*EnumMRUListFn)(HANDLE hMRU, int nItem, void* lpData, UINT uLen);
-typedef int (*FreeMRUListFn)(HANDLE hMRU);
+        Save();
+    }
+    else
+    {
+        ParseJson();
+    }
+}
+
+void MRUListHandler::Save()
+{
+    json::JsonObject jsonData;
+
+    jsonData.SetNamedValue(c_maxMRUSize, json::value(size));
+    jsonData.SetNamedValue(c_insertionIdx, json::value(pushIdx));
+    jsonData.SetNamedValue(c_mruList, Serialize());
+
+    json::to_file(jsonFilePath, jsonData);
+}
+
+json::JsonArray MRUListHandler::Serialize()
+{
+    json::JsonArray searchMRU{};
+
+    std::wstring data{};
+    for (const std::wstring& item : items)
+    {
+        searchMRU.Append(json::value(item));
+    }
+
+    return searchMRU;
+}
+
+void MRUListHandler::MigrateFromRegistry()
+{
+    std::wstring searchListKeys = GetRegString(c_mruList, registryFilePath);
+    std::sort(std::begin(searchListKeys), std::end(searchListKeys));
+    for (const wchar_t& key : searchListKeys)
+    {
+        Push(GetRegString(std::wstring(1, key), registryFilePath));
+    }
+}
+
+void MRUListHandler::ParseJson()
+{
+    auto json = json::from_file(jsonFilePath);
+    if (json)
+    {
+        const json::JsonObject& jsonObject = json.value();
+        try
+        {
+            long oldSize{ size };
+            if (json::has(jsonObject, c_maxMRUSize, json::JsonValueType::Number))
+            {
+                oldSize = (long)jsonObject.GetNamedNumber(c_maxMRUSize);
+            }
+            long oldPushIdx{ 0 };
+            if (json::has(jsonObject, c_insertionIdx, json::JsonValueType::Number))
+            {
+                oldPushIdx = (long)jsonObject.GetNamedNumber(c_insertionIdx);
+            }
+            if (json::has(jsonObject, c_mruList, json::JsonValueType::Array))
+            {
+                auto jsonArray = jsonObject.GetNamedArray(c_mruList);
+                if (oldSize == size)
+                {
+                    for (uint32_t i = 0; i < jsonArray.Size(); ++i)
+                    {
+                        items[i] = std::wstring(jsonArray.GetStringAt(i));
+                    }
+                    pushIdx = oldPushIdx;

```

Comment:
If user changed json file for any reason, validation would be worth doing. Otherwise this index could be out of bound when `Push` was called

Reply:
Validated `pushIdx`.